File: legal-automation/backend/app/schemas/dsgvo.py
from datetime import datetime

from pydantic import BaseModel


class ProcessingRecordResponse(BaseModel):
    id: int
    name: str
    purpose: str
    legal_basis: str
    data_categories: list
    data_subjects: str | None
    recipients: str | None
    retention: str | None
    tom: str | None
    is_active: bool

    model_config = {"from_attributes": True}


class ProcessingRecordCreate(BaseModel):
    name: str
    purpose: str
    legal_basis: str
    data_categories: list[str] = []
    data_subjects: str | None = None
    recipients: str | None = None
    retention: str | None = None
    tom: str | None = None


class RetentionPolicyResponse(BaseModel):
    id: int
    name: str
    matter_type: str | None
    retention_years: int
    legal_basis: str | None
    is_active: bool

    model_config = {"from_attributes": True}


class RetentionPolicyCreate(BaseModel):
    name: str
    matter_type: str | None = None
    retention_years: int
    legal_basis: str | None = None


class ErasureEligibilityResponse(BaseModel):
    allowed: bool
    blocking_reasons: list[str]


class ErasureRequestCreate(BaseModel):
    client_id: int
    reason: str | None = None


class ErasureRequestResponse(BaseModel):
    id: int
    client_id: int
    requested_by_id: int
    requested_at: datetime
    status: str
    reason: str | None
    blocking_reasons: list | None
    decided_at: datetime | None
    executed_at: datetime | None
    certificate_path: str | None

    model_config = {"from_attributes": True}


class DataExportResponse(BaseModel):
    id: int
    client_id: int
    status: str
    token: str
    expires_at: datetime | None
    download_path: str | None = None

    model_config = {"from_attributes": True}


class AdminOverviewResponse(BaseModel):
    active_sessions: int
    locked_users: int
    users_total: int
    users_with_2fa: int
    open_erasure_requests: int
    blocked_erasure_requests: int
    matters_past_retention: int

File: legal-automation/frontend/src/lib/api/dsgvo.ts
import api from "./client";

export interface ProcessingRecord {
  id: number;
  name: string;
  purpose: string;
  legal_basis: string;
  data_categories: string[];
  data_subjects: string | null;
  recipients: string | null;
  retention: string | null;
  tom: string | null;
  is_active: boolean;
}

export interface ErasureEligibility {
  allowed: boolean;
  blocking_reasons: string[];
}

export interface ErasureRequest {
  id: number;
  client_id: number;
  requested_by_id: number;
  requested_at: string;
  status: string;
  reason: string | null;
  blocking_reasons: string[] | null;
  decided_at: string | null;
  executed_at: string | null;
  certificate_path: string | null;
}

export interface AdminOverview {
  active_sessions: number;
  locked_users: number;
  users_total: number;
  users_with_2fa: number;
  open_erasure_requests: number;
  blocked_erasure_requests: number;
  matters_past_retention: number;
}

export interface RetentionPolicy {
  id: number;
  name: string;
  matter_type: string | null;
  retention_years: number;
  legal_basis: string | null;
  is_active: boolean;
}

export const dsgvoApi = {
  vvt: () => api.get<ProcessingRecord[]>("/dsgvo/vvt").then((r) => r.data),
  retentionPolicies: () => api.get<RetentionPolicy[]>("/dsgvo/retention-policies").then((r) => r.data),
  overview: () => api.get<AdminOverview>("/dsgvo/admin/overview").then((r) => r.data),

  eligibility: (clientId: number) =>
    api.get<ErasureEligibility>(`/dsgvo/erasure-eligibility/${clientId}`).then((r) => r.data),
  listErasure: () => api.get<ErasureRequest[]>("/dsgvo/erasure-requests").then((r) => r.data),
  createErasure: (client_id: number, reason?: string) =>
    api.post<ErasureRequest>("/dsgvo/erasure-requests", { client_id, reason }).then((r) => r.data),
  executeErasure: (id: number) =>
    api.post<ErasureRequest>(`/dsgvo/erasure-requests/${id}/execute`, {}).then((r) => r.data),
  rejectErasure: (id: number) =>
    api.post<ErasureRequest>(`/dsgvo/erasure-requests/${id}/reject`, {}).then((r) => r.data),

  createExport: (clientId: number) =>
    api.post<{ id: number; token: string; download_path: string; expires_at: string }>(
      `/dsgvo/export/${clientId}`, {}
    ).then((r) => r.data),
};
